frontend/app/src/App.test.tsx

```diff
@@ -219,1619 +369,1738 @@ describe("App", () => {
+      expect(screen.queryByText("Here is some text")).not.toBeInTheDocument()
     })
 
-    afterEach(() => {
-      pushStateSpy.mockRestore()
-      window.history.pushState({}, "", "/")
+    it("doesn't call clearAppState if currentPageScriptHash doesn't change", async () => {
+      await makeAppWithElements()
+
+      sendForwardMessage("newSession", NEW_SESSION_JSON)
+
+      expect(screen.getByText("Here is some text")).toBeInTheDocument()
     })
 
-    it("can switch to the main page from a different page", () => {
-      window.history.replaceState({}, "", "/page2")
+    describe("page change URL handling", () => {
+      let pushStateSpy: any
 
-      const instance = wrapper.instance() as App
-      instance.handleNewSession(new NewSession(NEW_SESSION_JSON))
+      beforeEach(() => {
+        window.history.pushState({}, "", "/")
+        pushStateSpy = jest.spyOn(window.history, "pushState")
+      })
 
-      expect(window.history.pushState).toHaveBeenLastCalledWith({}, "", "/")
-    })
+      afterEach(() => {
+        pushStateSpy.mockRestore()
+        window.history.pushState({}, "", "/")
+      })
 
-    it("can switch to a non-main page", () => {
-      const instance = wrapper.instance() as App
-      const newSessionJson = cloneDeep(NEW_SESSION_JSON)
-      newSessionJson.appPages = [
-        { pageScriptHash: "page_script_hash", pageName: "streamlit_app" },
-        { pageScriptHash: "hash2", pageName: "page2" },
-      ]
-      newSessionJson.pageScriptHash = "hash2"
+      it("can switch to the main page from a different page", () => {
+        renderApp(getProps())
+        window.history.replaceState({}, "", "/page2")
 
-      instance.handleNewSession(new NewSession(newSessionJson))
+        sendForwardMessage("newSession", NEW_SESSION_JSON)
 
-      expect(window.history.pushState).toHaveBeenLastCalledWith(
-        {},
-        "",
-        "/page2"
-      )
-    })
+        expect(window.history.pushState).toHaveBeenLastCalledWith({}, "", "/")
+      })
 
-    it("retains the query string", () => {
-      window.history.pushState({}, "", "/?foo=bar")
+      it("can switch to a non-main page", () => {
+        renderApp(getProps())
+        sendForwardMessage("newSession", {
+          ...NEW_SESSION_JSON,
+          appPages: [
+            { pageScriptHash: "page_script_hash", pageName: "streamlit_app" },
+            { pageScriptHash: "hash2", pageName: "page2" },
+          ],
+          pageScriptHash: "hash2",
+        })
 
-      const instance = wrapper.instance() as App
-      instance.handleNewSession(new NewSession(NEW_SESSION_JSON))
+        expect(window.history.pushState).toHaveBeenLastCalledWith(
+          {},
+          "",
+          "/page2"
+        )
+      })
 
-      expect(window.history.pushState).toHaveBeenLastCalledWith(
-        {},
-        "",
-        "/?foo=bar"
-      )
-    })
+      it("retains the query string", () => {
```

Comment:
Guessing this test will need to be changed to NOT retaining query strings between pages